### PR TITLE
Client module refactor

### PR DESCRIPTION
**Pull request by ctanveer**
### Description

Refactor of Client Module

### Link to ticket

*Please add a link to the GitHub issue being addressed by this change.*

### Type

- [ ] Fix
- [ ] Feature
- [ ] Cleanup

### CODE REVIEW

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 1]
```diff
@@ -52,38 +54,95 @@ public function init(
...
-                'name' => $row['name'],
-                'street' => $row['street'],
-                'zip' => $row['zip'],
-                'city' => $row['city'],
-                'state' => $row['state'],
-                'country' => $row['country'],
-                'phone' => $row['phone'],
-                'internet' => $row['internet'],
-                'email' => $row['email'],
+            $clientValues = (object) [
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:135]: The clientService should already return an object/model. This mapping should not be needed anymore.

File: app/Domain/Clients/Controllers/NewClient.php [hunk 2]
```diff
@@ -10,35 +10,61 @@
...
+        /**
+         * get - display template and provide empty values object
+         */
+        public function get(): Response
         {
+            Auth::authOrRedirect([Roles::$owner, Roles::$admin], true);
+
+            //Only admins
+            if (Auth::userIsAtLeast(Roles::$admin)) {
+                $values = (object) [
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/NewClient.php:38]: Instead of creating the array/object manually you should use the newly create client model to prepare an empty object. `$client = app()->make(ClientModel::class);`

On reason to use models is so that we don't have to worry about array definitions. For example, if I want to add a new field I shouldn't have to look for arrays in controllers. I just add it to the model and it will be available everywhere

File: app/Domain/Clients/Controllers/NewClient.php [hunk 3]
```diff
@@ -49,23 +75,23 @@ public function run()
...
-                        'zip' => ($_POST['zip']),
-                        'city' => ($_POST['city']),
-                        'state' => ($_POST['state']),
-                        'country' => ($_POST['country']),
-                        'phone' => ($_POST['phone']),
-                        'internet' => ($_POST['internet']),
-                        'email' => ($_POST['email']),
+                
+                if (isset($params['save']) === true) {
+                    $values = (object) [
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/NewClient.php:80]: Same as above. Please create a model and then map to the model directly. We don't do that everywhere but you can take at the tickets model. In there we use the constructor to map arrays to the model.

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 4]
```diff
@@ -52,38 +54,95 @@ public function init(
...
+
+            $row = $this->clientService->get($id);
+
+            if ($row === false) {
+                $this->tpl->display('errors.error404');
+
+                return $this->tpl->display('errors.error403');
+            }
+
+            $clientValues = (object) [
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:76]: Not needed since the service returns an object already.

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 5]
```diff
@@ -52,38 +54,95 @@ public function init(
...
+                'state' => $row->state,
+                'country' => $row->country,
+                'phone' => $row->phone,
+                'internet' => $row->internet,
+                'email' => $row->email,
+            ];
+
+
+            if (empty($row) === false && Auth::userIsAtLeast(Roles::$admin)) {
+                $file = app()->make(FileRepository::class);
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:91]: Can you pass this in via the constructor?

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 6]
```diff
@@ -52,38 +54,95 @@ public function init(
...
+                $project = app()->make(ProjectRepository::class);
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:92]: We are already passing in the projectService through the constructor, can you use that instead of the repo?

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 7]
```diff
@@ -52,38 +54,95 @@ public function init(
...
+
+                if (session('userdata.role') == 'admin') {
+                    $this->tpl->assign('admin', true);
+                }
+
+                $this->tpl->assign('userClients', $this->clientService->getUserClients($id));
+                $this->tpl->assign('comments', $this->commentService->getComments('client', $id));
+                $this->tpl->assign('imgExtensions', ['jpg', 'jpeg', 'png', 'gif', 'psd', 'bmp', 'tif', 'thm', 'yuv']);
+                $this->tpl->assign('client', $clientValues);
+                $this->tpl->assign('users', app()->make(UserRepository::class));
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:102]: Let's pass in the user service here and make the call in the controller. Haven't seen the template yet but I assume there is a repo method call in the template which sould probably not be in there.

File: app/Domain/Clients/Controllers/ShowClient.php [hunk 8]
```diff
@@ -118,37 +177,37 @@ public function run()
...
-                        'street' => $_POST['street'],
-                        'zip' => $_POST['zip'],
-                        'city' => $_POST['city'],
-                        'state' => $_POST['state'],
-                        'country' => $_POST['country'],
-                        'phone' => $_POST['phone'],
-                        'internet' => $_POST['internet'],
-                        'email' => $_POST['email'],
+                if (isset($params['save']) === true) {
+                    $clientValues = (object) [
```
Comment (marcelfolaron) [inline app/Domain/Clients/Controllers/ShowClient.php:188]: Please map using the object. In fact since you already received the client object from the db above you can just use that and override the values with your post values., better yet would be to just do that in the service so that we can do validation in there.

File: app/Domain/Clients/Models/Clients.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        
+        public $phone;
+        
+        public $internet;
+        
+        public $email;
+        
+        public $numberOfProjects;
+        
+        public function __construct() {}
```
Comment (marcelfolaron) [inline app/Domain/Clients/Models/Clients.php:29]: You could refactor the constructor to take an array and then map the array to your client model. See the ticket model for example

File: app/Domain/Clients/Repositories/Clients.php [hunk 10]
```diff
@@ -65,13 +66,15 @@ public function getClient($id): array|false
...
             $stmn->execute();
+            $stmn->setFetchMode(PDO::FETCH_CLASS, ClientsModel::class);
             $row = $stmn->fetch();
+            
             $stmn->closeCursor();
 
-            if ($row !== false && count($row) > 0) {
-                $this->name = $row['name'];
+            if ($row !== false) {
+                $this->name = $row->name;
```
Comment (marcelfolaron) [inline app/Domain/Clients/Repositories/Clients.php:75]: The following 2 lines don't really do anything? You can just return $row.

File: app/Domain/Clients/Repositories/Clients.php [hunk 11]
```diff
@@ -136,8 +140,9 @@ public function isClient($values): bool
 			name = :name AND street = :street LIMIT 1';
 
             $stmn = $this->db->database->prepare($sql);
-            $stmn->bindValue(':name', $values['name'], PDO::PARAM_STR);
-            $stmn->bindValue(':street', $values['street'], PDO::PARAM_STR);
+
+            $stmn->bindValue(':name', $values->name, PDO::PARAM_STR);
+            $stmn->bindValue(':street', $values->street, PDO::PARAM_STR);
```
Comment (marcelfolaron) [inline app/Domain/Clients/Repositories/Clients.php:145]: I know this was there before but this is actually weird. Why are we getting client by street? I think isClient() using the name only is fine. Can you check if we are actually using this method anywhere? If not, let's just remove it.

File: app/Domain/Clients/Repositories/Clients.php [hunk 12]
```diff
@@ -179,7 +185,7 @@ public function getClientsUsers($clientId): false|array
         /**
          * addClient - add a client and postback test
          */
-        public function addClient(array $values): false|string
+        public function addClient(object $values): false|string
```
Comment (marcelfolaron) [inline app/Domain/Clients/Repositories/Clients.php:188]: lastInsertId from below should return an int instead of string. Please change from false|string to false|int

File: app/Domain/Clients/Templates/delClient.blade.php [hunk 13]
```diff
@@ -12,7 +12,7 @@
     <div class="pageicon"><span class="fa fa-address-book"></span></div>
     <div class="pagetitle">
         <h5>{{ __("label.administration") }}</h5>
-        <h1><?php echo sprintf($tpl->__('headline.delete_client'), $client['name']); ?></h1>
+        <h1><?php echo sprintf($tpl->__('headline.delete_client'), $client->name); ?></h1>
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/delClient.blade.php:15]: Can you replace this with a blade syntax? `{!! sprintf( __('headline.delete_client'), $tpl->escape($client->name)) !!}`

File: app/Domain/Clients/Templates/delClient.blade.php [hunk 14]
```diff
@@ -22,8 +22,8 @@
     <div class="maincontentinner">
 
         @displayNotification()
-
-        <h4 class="widget widgettitle">{{ __("subtitles.delete") }}</h4>
+        {{-- {{ __("subtitles.delete") }} --}}
+        <h4 class="widget widgettitle"><i class="fa fa-trash"></i> Delete</h4>
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/delClient.blade.php:26]: Why was this changed? This hard codes the english string rather than using the translation. If it's the html being escaped please use `{!! __("subtitles.delete") !!}`

File: app/Domain/Clients/Templates/delClient.blade.php [hunk 15]
```diff
@@ -34,8 +34,7 @@
                 <x-global::forms.button type="submit" name="del" class="button">
                     {{ __('buttons.yes_delete') }}
                 </x-global::forms.button>
-
-                <x-global::forms.button tag="a" href="/clients/showClient/{{ $client['id'] }}" class="btn btn-primary">
+                <x-global::forms.button tag="a" href="/clients/showClient/{{ $client->id }}" class="btn btn-primary">
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/delClient.blade.php:37]: We should not need the class definitions here since they are already part of the component.

Comment (marcelfolaron) [reply app/Domain/Clients/Templates/delClient.blade.php:37]: While we're at it, can you change the "class="button" from above as well. 
Then lastly the secondary button should have content-role="tertiary"

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 16]
```diff
@@ -38,7 +38,7 @@
                                     inputType="text"
                                     name="name"
                                     id="name"
-                                    value="{{ $tpl->escape($values['name']) }}"
+                                    value="{{ $tpl->escape($values->name) }}"
                                     labelText="{{ __('label.name') }}"
                                     class="form-control"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:43]: Let's take out the form-control class from all of these

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 17]
```diff
@@ -38,7 +38,7 @@
                                     inputType="text"
                                     name="name"
                                     id="name"
-                                    value="{{ $tpl->escape($values['name']) }}"
+                                    value="{{ $tpl->escape($values->name) }}"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:41]: {{ }} is already escaping so you can just use the value directly {{ $values->name }}

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 18]
```diff
@@ -49,7 +49,7 @@ class="form-control"
                                     inputType="text"
                                     name="email"
                                     id="email"
-                                    value="{{ $tpl->escape($values['email']) }}"
+                                    value="{{ $tpl->escape($values->email) }}"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:52]: See above, remove escape

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 19]
```diff
@@ -60,7 +60,7 @@ class="form-control"
                                     inputType="text"
                                     name="internet"
                                     id="internet"
-                                    value="{{ $tpl->escape($values['internet']) }}"
+                                    value="{{ $tpl->escape($values->internet) }}"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:63]: See above, remove escape

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 20]
```diff
@@ -135,15 +135,24 @@ class="form-control"
...
 
                             <div class="form-group">
-                                <div class="span4 control-label">
+                                <div class="mt-4 span4 control-label">
                                     <x-global::forms.button
                                         type="submit"
                                         name="save"
+                                        value="true"
                                         id="save"
                                         class="btn btn-primary"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:144]: Please remove the btn classes from here

File: app/Domain/Clients/Templates/newClient.blade.php [hunk 21]
```diff
@@ -135,15 +135,24 @@ class="form-control"
...
                                     >
                                         {{ __('buttons.save') }}
                                     </x-global::forms.button>
+
+                                    <x-global::forms.button 
+                                        tag="a" 
+                                        href="/clients/showAll" 
+                                        class="btn btn-secondary"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/newClient.blade.php:152]: Remove btn class and set content-role

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 22]
```diff
@@ -13,7 +13,7 @@
     <div class="pageicon"><span class="fa fa-address-book"></span></div>
     <div class="pagetitle">
         <h5>{{ __("label.administration") }}</h5>
-        <h1><?php $tpl->e($values['name']); ?></h1>
+        <h1><?php $tpl->e($values->name); ?></h1>
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:16]: Please use blade syntax {{ $values->name }}

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 23]
```diff
@@ -45,7 +45,7 @@
                                     inputType="text" 
                                     name="id" 
                                     id="id" 
-                                    value="{{ $tpl->escape($values['id']) }}" 
+                                    value="{{ $tpl->escape($values->id) }}"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:48]: remove escape

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 24]
```diff
@@ -136,14 +136,14 @@
...
-                                    value="{{ $tpl->escape($values['phone']) }}" 
+                                    value="{{ $tpl->escape($values->phone) }}" 
                                     labelText="{{ __('label.phone') }}" 
                                 />
                             </div>
                         </div>
                                                 <div class="col-md-6">
                             <h4 class="widgettitle title-light"><span class="fa fa-users"></span> {{ __("subtitles.users_assigned_to_this_client") }}</h4>
-                            <a href="#/users/newUser?preSelectedClient=<?=$values['id'] ?>" class="btn btn-primary"><i class='fa fa-plus'></i> <?=$tpl->__('buttons.add_user') ?> </a>
+                            <a href="#/users/newUser?preSelectedClient=<?=$values->id ?>" class="btn btn-primary"><i class='fa fa-plus'></i> <?=$tpl->__('buttons.add_user') ?> </a>
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:146]: Please use blade syntax for the id and the add user strings

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 25]
```diff
@@ -178,14 +178,22 @@
...
-                            <x-global::forms.button type="submit" name="save" id="save">
+                            <x-global::forms.button type="submit" name="save" value="true" id="save">
                                 {{ __('buttons.save') }}
-                            </x-global::forms.button>                            
+                            </x-global::forms.button>
+                            
+                            <x-global::forms.button 
+                                tag="a" 
+                                href="/clients/showAll" 
+                                class="btn btn-secondary"
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:190]: Remove btn classes

Comment (marcelfolaron) [reply app/Domain/Clients/Templates/showClient.blade.php:190]: And set content-role

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 26]
```diff
@@ -178,14 +178,22 @@
...
+                            >
+                                {{ __('buttons.back') }}
+                            </x-global::forms.button>
                         </div>
                         <div class="col-md-6 align-right">
-                            <a href="{{ BASE_URL }}/clients/delClient/<?php $tpl->e($_GET['id']); ?>" class="delete"><i class="fa fa-trash"></i> {{ __("links.delete") }}</a>
+                            <a href="{{ BASE_URL }}/clients/delClient/<?php $tpl->e($values->id); ?>" class="delete"><i class="fa fa-trash"></i> {{ __("links.delete") }}</a>
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:196]: blade syntax

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 27]
```diff
@@ -194,9 +202,9 @@
 
                 <div id='comment'>
 
-                <form method="post" action="{{ BASE_URL }}/clients/showClient/<?php echo $tpl->e($_GET['id']); ?>#comment">
+                <form method="post" value="true" action="{{ BASE_URL }}/clients/showClient/<?php echo $tpl->e($values->id); ?>#comment">
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:205]: What does value="true" do in here? I don't think form has a "value" attribute

File: app/Domain/Clients/Templates/showClient.blade.php [hunk 28]
```diff
@@ -194,9 +202,9 @@
...
                     <input type="hidden" name="comment" value="1" />
-                    @include("comments::includes.generalComment", ["formUrl" => BASE_URL . "/clients/showClient/" . $tpl->escape($_GET['id'])])
+                    <x-comments::list :module="'client'" :statusUpdates="'true'" moduleId="{{ $tpl->escape($values->id) }}" />
```
Comment (marcelfolaron) [inline app/Domain/Clients/Templates/showClient.blade.php:207]: Can remove the escape

**Review by marcelfolaron — CHANGES_REQUESTED** [review]
Really good job overall. Small requests across the board

**Comment by ctanveer** [comment]
requested changes made

File: app/Domain/Clients/Templates/showClient.blade.php (hunk 27 above)
Comment (ctanveer) [reply app/Domain/Clients/Templates/showClient.blade.php]: without the value, the save property of array is set to null. In that case, the if condition in controller does not run

**Review by marcelfolaron — APPROVED** [review]
